### DISCUSSION TITLE
use of "->" operator (selector call expressions) for proc(s) with first parameter "this"

### DISCUSSION THREAD

**Discussion by renlite** [odin-lang/Odin#4316] [Proposals]
At the moment there are 3 steps necessary to use the syntax sugar of selector call expressions  "->":
1) declaration in the struct for type proc
2) define the implementation of the proc outside the struct
3) set the proc as value of the declaration.

See example below:

```Odin

Test :: struct {
	Name: string,
	// declaration hello
	_hello: proc(this: ^Test)   // 1)
}

//implementation hello
hello :: proc(this: ^Test) {    // 2)
	fmt.println(this.Name)
}

new_Test :: proc(name: string) -> Test {
	return Test{Name = name, _hello = hello}    // 3)
}

main :: proc() {

	foo_test := new_Test(name = "Foo")
	bar_test := Test{Name = "Bar"}
	hello(&foo_test)
	foo_test._hello(&foo_test)

	//*** -> operator (selector call expressions)
	foo_test->_hello()
	// No error - no output
	bar_test->_hello()
	
}

```` 
Solution 1):
Allow default values in the struct for all types. This would reduce Pt. 3) above.

```Odin

Test :: struct {
	Name: string,
	// declaration hello
	_hello: proc(this: ^Test) = hello   // 1)
}

//implementation hello
hello :: proc(this: ^Test) {    // 2)
	fmt.println(this.Name)
}

main :: proc() {
	bar_test := Test{Name = "Bar"}
	
	//*** -> operator (selector call expressions)
	bar_test->_hello()
	
}

```

Solution 2):
Use the definition/implementation of the proc outside a struct as a member ot the sruct, if the first parameter in the proc is a pointer to a struct in the package . This would reduce Pt. 1 and Pt. 3.

```Odin

Test :: struct {
	Name: string,
}

//implementation hello
hello :: proc(this: ^Test) {   // 2)
	fmt.println(this.Name)
}

main :: proc() {
	bar_test := Test{Name = "Bar"}
	
	//*** -> operator (selector call expressions)
	bar_test->hello()
	
}
```
I think Solution 2) would offer benefits without to break consistency and it is not an typical OO-design. You can still use the proc in common way. The first parameter name "this" could be mandatory if it fits design of Odin.

**Comment by flysand7** [comment]
You first described the way you might currently use selector call expressions, and then skipped directly to describing "solutions". Can you describe what the problem it is that you're trying to "solve"?

**Reply by gingerBill** [reply]
Exactly. What exactly is the problem?

**Reply by renlite** [reply]
Reduce the steps (less coding) to use sugar of selector call expressions "->".  Solution 1) wourd reduce the step 3 if default proc is ok. Methods style call is available -> it could be easier to use. Less to init if default values of properties (not only procs) are possible.

**Reply by captkirk88** [reply]
Curious why this hasn't been done.  All these arguments against it are specific.  Methods promote code structure.  Interfaces do too.  If you don't have methods you risk the abysmal `update_expiration_date_based_on_last_payment(...)` and that is the exact direction Odin and even Jai are going.  None of you have ever seen a long C function?  I would rather be able to call `PaymentExpiration.update(&lastPayment)` than have to create long proc names.

**Comment by laytan** [comment]
`->` is designed for objective-c and windows COM interfaces, not for what you do here. We don't want methods.

**Reply by renlite** [reply]
With "selector call expressions" you have already methods. You mentioned two reasons and further could follow. (UI FW?)
Default values in ```struct```s would offer the same benefits as in ```proc```s.

**Reply by perry2Erip** [reply]
I think having `x->y(foo, bar, etc)` as an alias for `y(x, foo, bar, etc)` wouldn't break Odin's philosophy if it just called `y` with the first parameter being `x` no matter what `x` is (as long as `y`'s first parameter accepts `x`'s type). It's more characters but language servers could use it to filter available completions.

**Reply by nmsobri** [reply]
i want methods

**Comment by tymbaca** [comment]
IMO the "only functions" approach makes the thought process simpler. You just don't need to think "do I need to make it a method or a function?" 

As an alternative, it can be a good feature for odin lsp (`ols`) to transform this `bar_test.hello()` to this `hello(bar_test)`.

**Reply by laytan** [reply]
OLS actually has this, if you put `"enable_fake_methods": true` in your config it will complete check for available functions of which you can pass the thing as the first parameter.

```odin

Foo :: struct {}

foo_hello :: proc(a: Foo) {}
foo_bar :: proc(a: ^Foo) {}

main :: proc() {
    f: Foo
    f.
}
```

Upon typing that period it would complete both of those procs and transform into the proper `foo_hello(f)` when accepted.

**Reply by flysand7** [reply]
big

**Reply by jabcross** [reply]
That kind of defeats the point of having left-to-right readability. The entire point of MCS is so that you don't have to navigate long nested parentheses or name every single partial result.

**Comment by captkirk88** [comment]
Here is a argument for struct methods.  Proc chaining.  I have never written a compiler in my life but I imagine it would help with optimizations internally where the compiler could just inline the chained procs.  It would also help me streamline a very small part of a api I'm working on.

```odin
MyStruct :: struct {}

MyStruct.doStuff :: proc() { return this }

MyStruct.doMoreStuff :: proc() { return this }

allmine: MyStruct = {}
allmine.doStuff().doMoreStuff()
```

@gingerBill I have heard on youtube and read your statement on methods and how much you hate them but there are usecases where they can help produce better code, proc chaining is a example of this.

**Reply by jabcross** [reply]
Can't you already do that with `->` ?

**Reply by antisaling** [reply]
@jabcross `->` would be akin to a virtual method so no optimization there (at least on the Odin frontend).

**Reply by antisaling** [reply]
@captkirk88 there is absolutely no difference between a concrete method call and a function call in terms of optimizations. I do understand the appeal of doing "chaining" but we eventually end up in a place where there are 3000 helpers for everything (some combined and some flat) because you either have to return the "this" or create a new thing. just see the current state of Rust where everyone has their favorite methods and the main point of contention on the language is that because structs implement things you now need to specify interfaces (`trait`s) on all compatible functions (impl<`T: MyTrait>`) exploding the polymorphism combinatorics which in turn explode compile time.

You are also inventing lots of C++-style syntax on this request which show no care for the existing Odin codestyle.

I do hope that you can understand this point to put a nail on the coffin of this request: it adds lots of complexity to the language and ends up being a footgun to your "optimization" conjecture.

**Reply by captkirk88** [reply]
@antisaling It is appealing to industry standards.  Chaining is in just about most industry adopted programming languages.  Is Odin intended to just be a hobby programming language then?  If so, then ignore this whole discussion and let's see how long this niche language lasts.  I agree on the part about Rust and I hate that language with a passion, however, I was referring to chaining which definitely has appeal.

I would suggest an alternative form of chaining then.. that fits this codestyle.

```odin
// Syntax: receiver .> procedure(args)
// Translated to: procedure(receiver, args)

res := data .> process_a() .> process_b(custom_arg)
```

The .> pipe-dot operator to signify that the left side is the first argument to the right side, only for procedures.  This does not turn the type into an object.  It remains a simple procedural pass.

Let me know what you think of this.

**Reply by jabcross** [reply]
[deleted and moved to standalone comment]

**Reply by jabcross** [reply]
> I do hope that you can understand this point to put a nail on the coffin of this request: it adds lots of complexity to the language and ends up being a footgun to your "optimization" conjecture.

I fail to see the implementation complexity, it could be a simple rewrite pass? It's just sugar, it doesn't even leave the AST... especially if a different symbol is used. Odin syntax already hides indirection by allowing MCS through `.` on pointers, I don't see why this is so different.

**Reply by antisaling** [reply]
@captkirk88 Chaining being common does not establish that Odin needs another spelling for procedure calls. C is an obvious counterexample (not that I'm defending it...), and language adoption depends far more on problem fit, tooling, ecosystem, and stability. My biggest 'fear' for this feature is that it moves Odin outside of its philosophical foundations on data-orientation, and trying to make a feature that COUPLES data with functionality which is basically dipping toes on OOP territory just because you want to write code in a different way.

**Reply by antisaling** [reply]
@jabcross @captkirk88  here are some points on complexity for you to think about.

1. you now need to add lots of syntax to handle instance, class, and static methods. Does the method always apply to an instance?
    ```odin
    Payment_Expiration.update :: proc(self: ^Payment_Expiration)
    Payment_Expiration.new :: proc(...) -> Payment_Expiration
    Payment_Expiration.days_remaining :: proc(...) -> int
    ```
    If so, you want syntax sugar so you do not have to write `self: ^Payment_Expiration` every time. Otherwise, you need separate syntax for constructors and static methods. Does `Payment_Expiration.new()` return an initialized value? Does `payment.new()` make sense? This is now secretly making Odin an OOP language.

2. how exactly do you declare an interface i.e. differentiate between methods and standard function-pointers as supported in the current implementation?

    ```odin
    Payment_Processor :: struct {
        process: proc(data: rawptr, payment: Payment) -> Error,
    }
    ```
    
    This is already a valid way to represent dynamic behavior. How is it different from:

    ```odin
    process :: proc(self: ^Payment_Processor, payment: Payment) -> Error
    ```
    
    Do methods always have to be declared inside the struct? If so, now structs have a second kind of member. Do methods have to be listed in an interface separately? If so, now the compiler needs method sets and interface satisfaction rules.
    
    So this goes into my previous comment that to be able to use this properly you also need something like a `trait`, or worse: Golang's `interface` thing.

4. what exactly is the receiver semantics? If the compiler hides the first parameter:

    ```odin
    recalculate :: proc(self: ^Invoice)
    invoice.recalculate()
    ```
    
    does this become `recalculate(&invoice)` automatically? What about:
    
    ```odin
    make_invoice().recalculate()
    ```
    
    Is the temporary addressable? Does the compiler copy it? What about value receivers, pointer receivers, nested pointers, map values, and fields returned from another procedure?
    
    Ordinary procedures make all of this explicit. Methods turn it into implicit language rules.

5. how do methods interact with `using`?

    ```odin
    Account :: struct {
        using metadata: Account_Metadata,
    }
    ```
    
    If `Account_Metadata` has a `touch` method, should this work?
    
    ```odin
    account.touch()
    ```
    
    If yes, the compiler needs to find the method through `metadata` and adjust the receiver. If two `using` fields provide the same method, it needs ambiguity rules. If methods are not promoted, then fields and methods behave differently for no obvious reason.

    Also note that this is secretly becoming actual inheritance from this point and there's no going back.

6. what does taking a method value mean?

    ```odin
    expire_payment := payment.expire
    expire_payment()
    ```
    
    Is `expire_payment` a bound `proc()` containing `payment`, or an unbound `proc(^Payment)` that still needs a receiver?
    
    If it is bound, the compiler needs storage for the receiver. This starts looking like a closure and introduces rules for copying it, mutating the receiver, escaping it, and representing it in the ABI.
    
    To be clear: I'm not particularly against closures (I do in fact have them implemented on my fork LOL). But this adds another "milestone" for a feature that you're assuming is simple.

7. how do generic methods specialize?

    ```odin
    Ring_Buffer :: struct($T: typeid) {
        data: []T,
    }
    ```
    
    If `Ring_Buffer(int)` has a `clear` method, does the compiler specialize `clear` from the receiver type?
    
    ```odin
    buffer.clear()
    ```
    
    What happens for `Ring_Buffer(string)`, pointers, aliases, nested generic types, or polymorphic receivers? The receiver is now part of generic inference and specialization identity instead of being an ordinary explicit argument.
    
    This is also changing how the compiler infers types. Making it a hella lot more implicit which Ginger is explicitly against

8. how do methods interact with fields and procedure groups?

    ```odin
    Document :: struct {
        save: proc() -> Error,
    }
    
    save :: proc(document: ^Document) -> Error
    ```
    
    What does this mean?
    
    ```odin
    document.save()
    ```
    
    Does the field win? Does the method win? Are both candidates? What if the field is an indirect callback but the method is a concrete procedure?
    
    Odin already has explicit procedure groups. Methods add implicit procedure groups attached to types.

9. how do methods interact with package ownership and foreign types? Can another package add a method to a type it did not declare?

    ```odin
    billing.Payment
    ```
    
    Can another package define:
    
    ```odin
    refund :: proc(payment: ^billing.Payment) -> Error
    ```
    
    and make this valid?
    
    ```odin
    payment.refund()
    ```
    
    If yes, method lookup depends on every package that imported the type. If no, the language needs ownership rules for aliases, named types, foreign declarations, and generated generic types.

    Again, the only sensible way of doing this would be with Rust-style `impl` blocks but now you are basically introducing additional namespacing into the language for no reason I guess.


All of this ties back into my previous argument: this is secretly making Odin an OOP language. Once values own methods, people naturally expect constructors, interfaces, private fields, inheritance, virtual dispatch, and lifecycle rules.

If only some of these are added, Odin gets the complexity of OOP without the complete model. If all of them are added, Odin is simply becoming an OOP language.

**Reply by jabcross** [reply]
I'll just focus on the `|>` proposal.

**Comment by jabcross** [comment]
(Originally an answer to the above thread but I think it stands on its own)

I think the main appeal of MCS is a structural one. A programming language is a way to project a control flow graph into a 1-d representation. Because that's topologically impossible for arbitrary graphs, we have identifiers (so that we can "jump" edges). In principle, with just identifiers and brackets we could represent any graph (as Lisp does), but having them as the only option takes away expressiveness in the form of the "rotation" of this 1-d projection.

In most programming languages we can do `1 + 2 - 3`, as well as `sub(add(1,2), 3)`. The difference is cosmetic, but I don't think anyone will disagree the first one looks much more readable. In the 2nd one, the `sub` is far away from its argument (the 3); the bigger the computation graph, the more difficult it becomes to read (and to keep track of the brackets).

Same in Python-style: `map(filter(map(x,f1),f2),f3)` is hard to read, as the arguments are far away from the proc call and the order of operations is mixed: calls are right-to-left, and arguments are left-to-right. (And this applies to a lot of design patterns, not just functional ones). Rewriting this as `map(x,f1) |> filter (f2) |> map(f3)` or `map(x,f1)->filter(f2)->map(f3)` or `map(x,f1).filter(f2).map(f3)` is just a rotation that does not change the original control flow graph at all, but makes it easier to read. Without this, we are forced to accept either the parentheses and semantic non-locality, or to name each partial result:

```odin
result1 = map(...)
result2 = filter(result1,...)
result3 = map(result2,...)
```

which I honestly find inelegant. Good naming is harder than people think, and it should only be necessary for self-documentation or for unavoidable structural limitations. More code is more places for bugs to hide in, and it incentivizes people to be lazy with their variable names.

(I know that the order of arguments of the actual functional package in Python is different, but the point stands for the argument. Selector calls in Odin already follow the convention of passing the first parameter through, and the issue of bracket nesting remains either way)

Alternatives:

- allow static/const struct members (uses struct namespace but not storage)

```odin
X :: struct {
   ...
   m :: proc (t: X, ...) ...
}
...

x := new(X)
X.m(x, ...)
x->m(...) // analogous to if `m` was a regular member of `X`
```
(overcomplicated IMO, namespaces are a rabbit hole and we're not remaking C++ here)

- use variable name shadowing (Rust's approach) to reuse variables, allowing to split the graph into several statements:

```odin
rv := f1(...)
rv := f2(rv,...)
rv := f3(rv,...)
```
(not a big fan of this one, shadowing is dangerous and there's still code repetition and boilerplate. also relies on the compiler optimizing out the extra assignments)

- Haskell-style backticks to make any function work as an operator (very neat but unconventional)

```odin
x `map` f1 `filter` f2 `map` f3 
```
(not really consistent with odins style)

---

I don't really see any way that solves as many of the aforementioned problems with as little change to the language. Just add a `|>` operator or similar, left-binding, lowest precedence, and have it rewrite `<begining_of_expr> |> <func_name>(<arg_list>) <rest_of_expr> ` to `<func_name>(<beginning_of_expr>,<arg_list>) <rest_of_expr>`.  

Only issue is resolving what we're passing through if the incoming expression is a function call with multiple results. (maybe `|>` passes all of them, and `?>` passes the first or short-circuits the entire pipeline if the second one is not nil, keeping compatibility with the ok idiom like an expression-level `or_return`, or an `or_else` that passes through the first result.)

**Reply by antisaling** [reply]
Your control-flow-graph framing is basically conflating source notation, expression structure, data flow, and control flow.

`map(filter(map(x, f1), f2), f3)` is just too trivial an example, see that a pipeline does not solve the general problem of representing arbitrary graphs in one-dimensional text. Once you have any kind of control flow or simply reading/reusing the second return of something names and statements become necessary again. And you also have the onus of rewriting the whole thing in a different style so this doesn't really favor code reuse either.

There is also an API-design cost. A universal pipeline privileges the first parameter encouraging procedures to place the “flowing” value first, even when an allocator, destination, context, or another operand would otherwise be the natural first parameter. APIs that do not follow this convention either cannot participate cleanly or acquire wrappers created primarily for chaining. So now this influences API design in a weird way.

Also note that arguing this
```
angle :=
    vector_a
    |> normalize()
    |> dot(vector_b |> normalize())
    |> clamp(-1, 1)
    |> arccos()
```
somehow reads better than this
```odin
angle := arccos(
    clamp(
        dot(
            normalize(vector_a),
            normalize(vector_b),
        ),
        -1,
        1,
    ),
)
```
is a VERY personal opinion... it took me good 10min to get your syntax right and even now I struggle to read it.

I also disagree that named intermediate values are merely boilerplate imposed by structural limitations. Names can expose semantic stages, types, ownership, mutation, error boundaries, and useful debugger locations:

```odin
first_derivative := differentiate(polynomial)
critical_points := real_roots(first_derivative) or_return
second_derivative := differentiate(first_derivative)
classified_points := classify_critical_points(second_derivative, critical_points)
```

Those names communicate information that a pipeline removes. “More lines” does not automatically mean “more places for bugs.” Semantic density matters more than textual line count. In some sense your proposal could backfire when you're trying to debug.

Taking Rust or C++ as examples for MCS doesn't really suit Odin's core design because they are tied to language protocols e.g.

- Rust: `Iterator`, `Option`, and `Result`.
- C++: member chaining, builders, and range-adaptor pipelines.

Also note that neither one establishes that every arbitrary free-function call benefits from a universal first-argument pipeline. They establish that certain linear APIs benefit from chain-oriented design.

It is also unclear whether the “short-circuit” operator returns from the enclosing procedure (producing an expression fallback) or behaving like `or_else`. Those are different operations. So the main issue I see is that we could read it in 4 ways and none of them are clear from syntax alone (tbh I already find `or_return` a bit of a pain to use correctly). To illustrate consider `v := x |> parse()?> normalize()` and suppose parse returns (value, err). If err != nil, what happens?
1. Return err from enclosing procedure, like or_return.
2. Stop pipeline but somehow produce value for surrounding expression.
3. Use fallback value, like or_else.
4. Return both values without running remaining stages.

So now can `v` be typed as something different depending on how far you went? If you want to actually propose something like `|>` I would leave `?>` out of it for sure.

As a final argument note that you're simply introducing another kind of non-locality with this syntax because now the arguments for a function are not "there" anymore. e.g. going back to your arithmetic example: `1 |> add(2) |> sub(3)` reads kind of arbitrary because the two values are symmetric.


---
I agree that pipeline-notation may be clearer when computation is strictly linear and intermediate stages have no useful semantic identity. I just think you need to provide better examples of use-intent because right now this seems like a niche thing. And please don't assume this doesn't have implications for the language.

**Reply by jabcross** [reply]
> Your control-flow-graph framing is basically conflating source notation, expression structure, data flow, and control flow.
> 
> Once you have any kind of control flow or simply reading/reusing the second return of something names and statements become necessary again.

The "second return" thing is kinda unique to Odin, but we can still deal with it with Maybe() and such. Even so, there are solutions to these problems, and even if not, just the single-return case already earns its keep IMO.

> And you also have the onus of rewriting the whole thing in a different style 

That's not an onus, that's flexibility. You don't *have* to rewrite everything, you have the option to.

> the natural first parameter

What makes a parameter a "natural" first parameter? And Odin already has this behavior with `->`. All I want is to be able to use something similar without needing a useless vtable when I already know every function pointer at compile time. 

> So now this influences API design in a weird way.

Nothing weird about state-carrying structures. It's a very well established pattern, both in OOP and in functional languages.

```
IRState :: struct { ... }

ir := new(IRState)
 |> add_import()
 |> add_function(..., new(IRState)
    |> add_statement(...)
    |> add_return()
  )
 |> add_function(...) 
```

The obvious benefit here is that the indentation matches the hierarchy of the object being constructed. Can you imagine having to re-indent everything just from adding a new item to the middle of the list?

Also, the API adaptation problem is something that's usually solved with compact lambdas, but that's a different discussion.

> is a VERY personal opinion... it took me good 10min to get your syntax right and even now I struggle to read it.

So you're saying that the level of nesting needs to be proportional to the length of the pipeline? Also, if you want to add an additional step in the middle of the pipeline, instead of a 1-line git diff (just add a line with a `|>`) you will have to change the indentation and add a pair of brackets to the *entire* pipeline. It's the semantic locality thing. To me, this by itself justifies the whole thing.

Also, not very personal, it's standard functional style. Elixir and Haskell code look like this. (I do think it reads better just from the fact that time flows from top to bottom. Don't have to go up and down and count indentations to find out where you are. Indentation represents converging dataflow edges rather than how many steps you've completed in the pipeline.)

> Those names communicate information that a pipeline removes. “More lines” does not automatically mean “more places for bugs.” Semantic density matters more than textual line count. In some sense your proposal could backfire when you're trying to debug.

I know. I explicitly mentioned self-documentation. I'm saying it should be up to the user. We can already nest things to avoid the naming if we want to, this doesn't change anything. (In fact, the pipeline style is _easier_ than the nested style when editing to add variable names if needed.) And, pipeline style is semantically denser than name-every-step style. No redundancy from repeated variable names.

Also, your example is not really what I was talking about. I'm talking for instance about type-tetris kind of situations where we have a sequence of type transformations that all refer to the same data in different forms (such as converting arrays to slices, etc). You want `data_file_name`, `data_file_descriptor`, `data_array`, `data_slice`, `data_ptr`, etc when all this information is obvious from the function calls? As I said, it's an incentive for lazy naming.


> Taking Rust or C++ as examples for MCS doesn't really suit Odin's core design because they are tied to language protocols e.g.
>
>   Rust: Iterator, Option, and Result.
>   C++: member chaining, builders, and range-adaptor pipelines.

These are not language protocols, though? These are design patterns, they can be found in many languages. Member chaining and builders in particular are ubiquitous, not sure why you associate them with C++. Range-adaptor pipelines are C++ realizing that pipelines are useful, a decade after they became industry standard (they've been in Boost for a LONG time). Option and Result are explicitly mentioned *not* to be Rust special cases, they're just regular enums.

> Also note that neither one establishes that every arbitrary free-function call benefits from a universal first-argument pipeline. They establish that certain linear APIs benefit from chain-oriented design.

And why is this not a good enough argument? Type safety ensures that usage is unambiguous, it doesn't affect people who don't want to use pipelines at all.

> It is also unclear whether the “short-circuit” operator returns from the enclosing procedure (producing an expression fallback) or behaving like `or_else`. Those are different operations. So the main issue I see is that we could read it in 4 ways and none of them are clear from syntax alone (tbh I already find `or_return` a bit of a pain to use correctly). To illustrate consider `v := x |> parse()?> normalize()` and suppose parse returns (value, err). If err != nil, what happens?
> 
>     1. Return err from enclosing procedure, like or_return.
> 
>     2. Stop pipeline but somehow produce value for surrounding expression.
> 
>     3. Use fallback value, like or_else.
> 
>     4. Return both values without running remaining stages.
> 

I was thinking that the entire `x |> parse() ?> normalize()` expression would act equivalent to `x |> parse()` if `err != nil`. You capture err like everywhere else, in the assignment. ` v, err := ` get assigned to nil and the err from parse(). So, 4 I guess? But this is more Odin-specific, so it's something to be discussed.

> So now can `v` be typed as something different depending on how far you went? If you want to actually propose something like `|>` I would leave `?>` out of it for sure.

v would be typed to the first result of the last pipeline stage. It's nil or default-valued if it returns early, analogous to an `or_return`. Could capture the actual return of parse() if it was an `any`, though...

> As a final argument note that you're simply introducing another kind of non-locality with this syntax because now the arguments for a function are not "there" anymore. e.g. going back to your arithmetic example: `1 |> add(2) |> sub(3)` reads kind of arbitrary because the two values are symmetric.

The symmetry is specific to add, though. And what do you mean, the arguments aren't there? `|>` means 'first argument of next function call is everything to my left'. It's directly adjacent to the function call, what's more local than that? Again, not arbitrary, it's already very well established in several other languages.

> I agree that pipeline-notation may be clearer when computation is strictly linear and intermediate stages have no useful semantic identity. I just think you need to provide better examples of use-intent because right now this seems like a niche thing. And please don't assume this doesn't have implications for the language.

Still unclear on these implications; how will the existence of this feature affect any user who doesn't want to use it? `Maybe()` is in the language even though it's not used very often in the standard lib, I don't see how this is different. Odin seems accomodating for many use-cases, I was surprised that selector calls are even in the language. I just think it's a bit odd that vtables are easier to use than pipelines in Odin.

**Reply by captkirk88** [reply]
@gingerBill let's discuss.

Odin is on the verge of 1.0 and over the course of it's development, language features have become optional ->  `#+feature dynamic-literals using-stmt`...

**Reply by antisaling** [reply]
> I just think it's a bit odd that vtables are easier to use than pipelines in Odin.

Well because to talk to Darwin and other FFI-applications you need to have some little ergonomics, and interfaces are a VERY important feature that also makes use of that.
Pipelines OTOH are borrowing syntax sugar from functional languages (which Odin is not) to create equivalent syntax.

Keep in mind Odin IS a minimalistic systems language, not Haskell. but I won't argue anymore. gotta admit this is the least bad idea on this whole discussion.

**Reply by jabcross** [reply]
Fair enough, thanks for the compliment haha
